src/core/lib/promise/activity.h

```diff
@@ -85,19 +97,50 @@ class Waker {
   }
 
  private:
-  struct Unwakeable final : public Wakeable {
-    void Wakeup() override {}
-    void Drop() override {}
-  };
-  static Unwakeable* unwakeable() {
-    return NoDestructSingleton<Unwakeable>::Get();
-  }
+  friend class AtomicWaker;
 
-  Wakeable* Take() { return absl::exchange(wakeable_, unwakeable()); }
+  Wakeable* Take() {
+    return absl::exchange(wakeable_, activity_detail::unwakeable());
+  }
 
   Wakeable* wakeable_;
 };
 
+// An atomic variant of Waker - this type is non-copyable and non-movable.
+class AtomicWaker {
+ public:
+  explicit AtomicWaker(Wakeable* wakeable) : wakeable_(wakeable) {}
+  AtomicWaker() : AtomicWaker(activity_detail::unwakeable()) {}
+  explicit AtomicWaker(Waker waker) : AtomicWaker(waker.Take()) {}
+  ~AtomicWaker() { wakeable_.load(std::memory_order_relaxed)->Drop(); }
+  AtomicWaker(const AtomicWaker&) = delete;
+  AtomicWaker& operator=(const AtomicWaker&) = delete;
+  AtomicWaker(AtomicWaker&& other) noexcept = delete;
+  AtomicWaker& operator=(AtomicWaker&& other) noexcept = delete;
+
+  // Wake the underlying activity.
+  void Wakeup() { Take()->Wakeup(); }
+
+  // Return true if there is a not-unwakeable wakeable present.
+  bool Armed() const noexcept {
+    return wakeable_.load(std::memory_order_relaxed) !=
+           activity_detail::unwakeable();
+  }
+
+  // Set to some new waker
+  void Set(Waker waker) {
+    wakeable_.exchange(waker.Take(), std::memory_order_relaxed)->Drop();
```

Comment:
Should there be any preconditions asserted on what sort of waker is being overwritten? For example, if a caller expects it to be in an unwakeable state, but there's a race for setting the value, this method could overwrite some other thread's waker that was set just beforehand.

Reply:
I think I need some more experience with this before deciding what's good or bad there... but we should actually be defaulting to Wakeup here just in case, so fixed that.